File: MCTS_TicTacToe_python/Node.py
from random import randint
from State import State


class Node:
    def __init__(self):
        self.parent = None
        self.state = State()
        self.children = []

    def get_random_child_node(self):
        no_of_possible_moves = len(self.children)
        select_random = randint(0, no_of_possible_moves - 1)
        return self.children[select_random]

    def get_child_with_max_score(self):
        return max(self.children, key=lambda child: child.state.visits)

File: MCTS_BlokusDuo_java/src/com/joaomiguelvieira/Node.java
package com.joaomiguelvieira;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;

public class Node {
    private State state;
    private ArrayList<Node> childArray;
    private Node parent;

    public Node() {
        state = new State();
        childArray = new ArrayList<>();
    }

    public Node(State state) {
        this.state = state;
        childArray = new ArrayList<>();
    }

    public State getState() {
        return state;
    }

    public ArrayList<Node> getChildArray() {
        return childArray;
    }

    public void setParent(Node parent) {
        this.parent = parent;
    }

    public Node getRandomChildNode() {
        int noOfPossibleMoves = childArray.size();
        int selectRandom = (int) (Math.random() * noOfPossibleMoves);
        return childArray.get(selectRandom);
    }

    public Node getParent() {
        return parent;
    }

    public Node getChildWithMaxScore() {
        return Collections.max(childArray, Comparator.comparing(node -> node.getState().getStateViability()));
    }
}
